src/array.rs

```diff
@@ -0,0 +1,456 @@
+    pub fn from_float(val: f32) -> Array {
+        let ptr = unsafe { mlx_sys::mlx_array_from_float(val) };
+        Array { ptr }
+    }
+
+    /// New array from a complex scalar.
+    pub fn from_complex(val: c64) -> Array {
+        let ptr = unsafe { mlx_sys::mlx_array_from_complex(val.re, val.im) };
+        Array { ptr }
+    }
+
+    /// New array from existing buffer.
+    ///
+    /// # Parameters
+    ///
+    /// - `data`: A buffer which will be copied.
+    /// - `shape`: Shape of the array.
+    /// - `dim`: Number of dimensions (size of shape).
+    pub fn from_data<T: ArrayElement>(data: &[T], shape: &[i32], dim: i32) -> Self {
+        let ptr = unsafe {
+            mlx_sys::mlx_array_from_data(
+                data.as_ptr() as *const c_void,
+                shape.as_ptr(),
+                dim,
+                T::DTYPE as u32,
+            )
+        };
+
+        Array { ptr }
+    }
+
+    /// The size of the array’s datatype in bytes.
+    pub fn itemsize(&self) -> usize {
+        unsafe { mlx_sys::mlx_array_itemsize(self.ptr) }
+    }
+
+    /// Number of elements in the array.
+    pub fn size(&self) -> usize {
+        unsafe { mlx_sys::mlx_array_size(self.ptr) }
+    }
+
+    /// The strides of the array.
+    pub fn strides(&self) -> &[usize] {
+        let ndim = self.ndim();
+
+        unsafe {
+            let data = mlx_sys::mlx_array_strides(self.ptr);
+            std::slice::from_raw_parts(data, ndim)
+        }
+    }
+
+    /// The number of bytes in the array.
+    pub fn nbytes(&self) -> usize {
+        unsafe { mlx_sys::mlx_array_nbytes(self.ptr) }
+    }
+
+    /// The array’s dimension.
+    pub fn ndim(&self) -> usize {
+        unsafe { mlx_sys::mlx_array_ndim(self.ptr) }
+    }
+
+    /// The shape of the array.
+    ///
+    /// Returns: a pointer to the sizes of each dimension.
+    pub fn shape(&self) -> &[i32] {
+        let ndim = self.ndim();
+
+        unsafe {
+            let data = mlx_sys::mlx_array_shape(self.ptr);
+            std::slice::from_raw_parts(data, ndim)
+        }
+    }
+
+    /// The shape of the array in a particular dimension.
+    ///
+    /// # Panic
+    ///
+    /// Panics if the array is scalar.
+    pub fn dim(&self, dim: i32) -> i32 {
+        // This will panic on a scalar array
+        unsafe { mlx_sys::mlx_array_dim(self.ptr, dim) }
+    }
+
+    /// The array element type.
+    pub fn dtype(&self) -> Dtype {
+        let dtype = unsafe { mlx_sys::mlx_array_get_dtype(self.ptr) };
+        Dtype::try_from(dtype).unwrap()
+    }
+
+    /// Evaluate the array.
+    pub fn eval(&mut self) {
+        // This clearly modifies the array, so it should be mutable
+        unsafe { mlx_sys::mlx_array_eval(self.ptr) };
+    }
+
+    /// Access the value of a scalar array.
+    pub fn item<T: ArrayElement>(&self) -> T {
+        T::scalar_array_item(self)
+    }
+
+    /// Returns a pointer to the array data
+    pub fn data<T: ArrayElement>(&self) -> &[T] {
```

Comment:
I don't understand how this is used yet, so I can't comment greatly on what type we need to return here. Shouldn't this be a &[u8] if we're returning byte data? 

either way, let's leave it for now.. but! you want to make sure you call `self.eval()` at the beginning here.

Reply:
This is actually marked as for internal use (or use with caution) in the c++ code, so I was kinda surprised that this is exposed in the c api. 

> Shouldn't this be a &[u8] if we're returning byte data?

The [c++ code](https://github.com/ml-explore/mlx/blob/b18468bf81ded999fda3971289f712920ef0b777/mlx/array.h#L308) returns a `*T`. I think we should probably be consistent with that

> but! you want to make sure you call self.eval() at the beginning here.

That's why I had the TODO there. I want to come back after we have implemented the ops

Reply:
Should this be renamed to `as_slice`? The swift binding renamed it to `asArray`

Reply:
Just saw that swift has the `Data` type representing raw bytes. I guess renaming to `as_slice` and `from_slice` would help avoid the confusion

Reply:
The swift binding calls `eval` internally. We would have to make `as_slice` to take `&mut self` if we want to do the same. Or should we just document this and ask the user to call `eval()`?

Reply:
Nah, I think we should assume intent. My guess is that it'll be easier to forget that you need to call eval rather than doing it. We can always change it later.

Reply:
My concern is still data racing. Let me double check if the c++ side does anything in this regard for `eval()`

Reply:
The way mlx determines whether an `array` is `eval`ed is by checking whether the `array_desc_->data` ptr is set. The `array_desc_->data` ptr is a `std::shared_ptr` but not protected by `atomic`. Plus the usual impl of the primitives seem to allocate and set the `data` ptr before all output are written to the newly allocated buffer. So if we call `mlx_array_eval` implicitly within a fn that only immutably borrows, we could likely end up with data racing

Reply:
`as_slice` returns `None` if the array is not `eval`ed.